File: templates/typescript/plugins/CrewAIPlugin.ts
/**
 * CrewAI → Borgkit Plugin (TypeScript) — HTTP Bridge
 * ─────────────────────────────────────────────────────────────────────────────
 * Bridges a running CrewAI HTTP service into the Borgkit mesh.
 *
 * CrewAI is Python-native; this plugin calls it over HTTP so TypeScript agents
 * can invoke CrewAI crews without shipping Python bindings.
 *
 * ── Expected service endpoints ────────────────────────────────────────────────
 *
 *   GET  /capabilities
 *     → [{ name: string, description: string, parameters?: Record }]
 *
 *   POST /kickoff
 *     Body:     { capability?: string, inputs: Record, task?: string }
 *     Response: { result: string, status: "success"|"error", error?: string }
 *
 * ── Serving CrewAI over HTTP ──────────────────────────────────────────────────
 *
 * Option A — zero-dependency FastAPI wrapper (drop next to your crew):
 *
 *   # serve_crew.py
 *   from fastapi import FastAPI
 *   from my_crew import my_crew            # your Crew instance
 *
 *   app = FastAPI()
 *
 *   @app.get("/capabilities")
 *   def caps():
 *       return [{"name": "kickoff", "description": "Run the crew on a task"}]
 *
 *   @app.post("/kickoff")
 *   async def kickoff(body: dict):
 *       result = my_crew.kickoff(inputs=body.get("inputs", {}))
 *       return {"result": str(result), "status": "success"}
 *
 *   # uvicorn serve_crew:app --port 8000
 *
 * Option B — borgkit-crewai Python adapter (auto-generates endpoints):
 *
 *   pip install borgkit-crewai
 *   borgkit-crewai serve --crew my_package:my_crew --port 8000
 *
 * ── Usage ─────────────────────────────────────────────────────────────────────
 *
 *   import { CrewAIPlugin } from './plugins/CrewAIPlugin';
 *
 *   const plugin = new CrewAIPlugin({
 *     agentId:    'borgkit://agent/researcher',
 *     name:       'ResearchAgent',
 *     version:    '1.0.0',
 *     owner:      '0xYourWallet',
 *     serviceUrl: 'http://localhost:8000',
 *     tags:       ['research', 'crewai'],
 *   });
 *
 *   // Eagerly fetch capabilities from the service (optional but recommended):
 *   await plugin.fetchCapabilities();
 *
 *   const agent = plugin.wrap(plugin.createService());
 *   await agent.registerDiscovery();
 */

import { AgentRequest }        from '../interfaces/IAgentRequest';
import { AgentResponse }       from '../interfaces/IAgentResponse';
import { BorgkitPlugin, PluginConfig, CapabilityDescriptor } from './IPlugin';

// ── Config ────────────────────────────────────────────────────────────────────

export interface CrewAIPluginConfig extends PluginConfig {
  /** Base URL of the running CrewAI HTTP service (default: 'http://localhost:8000') */
  serviceUrl?: string;
  /** Optional Bearer token sent as Authorization header */
  apiKey?: string;
  /** HTTP request timeout in milliseconds (default: 120 000) */
  timeoutMs?: number;
  /**
   * Static capability list — skips GET /capabilities entirely.
   * Useful when the service does not expose a capabilities endpoint.
   */
  staticCapabilities?: Array<{
    name:         string;
    description:  string;
    parameters?:  Record<string, unknown>;
  }>;
  /** POST path for crew execution (default: '/kickoff') */
  kickoffPath?: string;
  /** GET path for capability discovery (default: '/capabilities') */
  capabilitiesPath?: string;
}

// ── Service handle ────────────────────────────────────────────────────────────

/**
 * Opaque token representing a live CrewAI HTTP service.
 * Create one via `plugin.createService()`.
 */
export interface CrewAIService {
  serviceUrl:       string;
  apiKey?:          string;
  timeoutMs:        number;
  kickoffPath:      string;
  capabilitiesPath: string;
}

// ── Wire types ────────────────────────────────────────────────────────────────

interface CrewAIInput {
  capability?: string;
  task?:       string;
  inputs:      Record<string, unknown>;
}

interface CrewAIOutput {
  result?:  string;
  output?:  string;
  status?:  string;
  error?:   string;
  [key: string]: unknown;
}

interface RemoteCapability {
  name:        string;
  description: string;
  parameters?: Record<string, unknown>;
}

// ── Plugin ────────────────────────────────────────────────────────────────────

export class CrewAIPlugin extends BorgkitPlugin<CrewAIService, CrewAIInput, CrewAIOutput> {
  private readonly crewConfig: {
    serviceUrl:       string;
    apiKey?:          string;
    timeoutMs:        number;
    kickoffPath:      string;
    capabilitiesPath: string;
    staticCapabilities?: CrewAIPluginConfig['staticCapabilities'];
  };

  /** Remote capabilities fetched via GET /capabilities (cached after first fetch). */
  private _remoteCapabilities?: CapabilityDescriptor[];

  constructor(config: CrewAIPluginConfig) {
    super(config);
    this.crewConfig = {
      serviceUrl:         config.serviceUrl        ?? 'http://localhost:8000',
      apiKey:             config.apiKey,
      timeoutMs:          config.timeoutMs          ?? 120_000,
      kickoffPath:        config.kickoffPath        ?? '/kickoff',
      capabilitiesPath:   config.capabilitiesPath   ?? '/capabilities',
      staticCapabilities: config.staticCapabilities,
    };
  }

  // ── Factory ────────────────────────────────────────────────────────────────

  /** Build the service handle to pass to `plugin.wrap()`. */
  createService(): CrewAIService {
    return {
      serviceUrl:       this.crewConfig.serviceUrl,
      apiKey:           this.crewConfig.apiKey,
      timeoutMs:        this.crewConfig.timeoutMs,
      kickoffPath:      this.crewConfig.kickoffPath,
      capabilitiesPath: this.crewConfig.capabilitiesPath,
    };
  }

  // ── Capability extraction ──────────────────────────────────────────────────

  /**
   * Resolve capabilities using this priority:
   *   1. `staticCapabilities` from config
   *   2. Cached capabilities from a prior `fetchCapabilities()` call
   *   3. Single `'invoke'` fallback (lazy — fetched on first HTTP call if needed)
   */
  extractCapabilities(_service: CrewAIService): CapabilityDescriptor[] {
    if (this.crewConfig.staticCapabilities?.length) {
      return this.crewConfig.staticCapabilities.map(c => this._toDescriptor(c));
    }
    if (this._remoteCapabilities?.length) {
      return this._remoteCapabilities;
    }
    return [{
      name:        'invoke',
      description: this.config.description ?? 'Invoke the CrewAI crew',
      nativeName:  '__crew__',
      tags:        this.config.tags ?? [],
    }];
  }

  /**
   * Eagerly fetch capabilities from the service's GET /capabilities endpoint.
   *
   * Call this before `plugin.wrap()` to populate the capability list so that
   * discovery announces all individual crew tools.
   *
   * @example
   * ```ts
   * await plugin.fetchCapabilities();
   * const agent = plugin.wrap(plugin.createService());
   * ```
   */
  async fetchCapabilities(): Promise<CapabilityDescriptor[]> {
    const svc = this.createService();
    const url = `${svc.serviceUrl.replace(/\/$/, '')}${svc.capabilitiesPath}`;
    const res = await this._doFetch(svc, url, 'GET');
    if (!res.ok) {
      throw new Error(`GET ${url} returned ${res.status}`);
    }
    const data = await res.json() as unknown;
    const list = Array.isArray(data) ? (data as RemoteCapability[]) : [];
    this._remoteCapabilities = list.map(c => this._toDescriptor(c));
    return this._remoteCapabilities;
  }

  private _toDescriptor(c: RemoteCapability): CapabilityDescriptor {
    return {
      name:        c.name,
      description: c.description,
      nativeName:  c.name,
      tags:        this.config.tags ?? [],
      inputSchema: c.parameters
        ? { type: 'object', properties: c.parameters }
        : undefined,
    };
  }

  // ── Request translation ────────────────────────────────────────────────────

  translateRequest(req: AgentRequest, descriptor: CapabilityDescriptor): CrewAIInput {
    const payload = req.payload as Record<string, unknown>;
    return {
      capability: descriptor.nativeName === '__crew__' ? undefined : descriptor.nativeName,
      task:       (payload['task'] ?? payload['query'] ?? payload['input']) as string | undefined,
      inputs:     payload,
    };
  }

  // ── Response translation ───────────────────────────────────────────────────

  translateResponse(output: CrewAIOutput, requestId: string): AgentResponse {
    if (output.status === 'error' || output.error) {
      return {
        requestId,
        status:       'error',
        errorMessage: output.error ?? JSON.stringify(output),
        timestamp:    Date.now(),
      };
    }
    const content = output.result ?? output.output ?? JSON.stringify(output);
    return {
      requestId,
      status:    'success',
      result:    { content, raw: output },
      timestamp: Date.now(),
    };
  }

  // ── Native invocation ──────────────────────────────────────────────────────

  async invokeNative(
    service:    CrewAIService,
    _d:         CapabilityDescriptor,
    input:      CrewAIInput,
  ): Promise<CrewAIOutput> {
    const url = `${service.serviceUrl.replace(/\/$/, '')}${service.kickoffPath}`;
    const res = await this._doFetch(service, url, 'POST', JSON.stringify(input));

    if (!res.ok) {
      const text = await res.text().catch(() => '');
      throw new Error(`CrewAI service returned HTTP ${res.status}: ${text}`);
    }

    return res.json() as Promise<CrewAIOutput>;
  }

  // ── HTTP helper ────────────────────────────────────────────────────────────

  private async _doFetch(
    service: CrewAIService,
    url:     string,
    method:  'GET' | 'POST',
    body?:   string,
  ): Promise<Response> {
    const headers: Record<string, string> = { Accept: 'application/json' };
    if (body)            headers['Content-Type'] = 'application/json';
    if (service.apiKey)  headers['Authorization'] = `Bearer ${service.apiKey}`;

    const controller = new AbortController();
    const timer      = setTimeout(() => controller.abort(), service.timeoutMs);

    try {
      const res = await fetch(url, { method, headers, body, signal: controller.signal });
      clearTimeout(timer);
      return res;
    } catch (err) {
      clearTimeout(timer);
      const isTimeout = err instanceof Error && err.name === 'AbortError';
      throw isTimeout
        ? new Error(`CrewAI service timed out after ${service.timeoutMs} ms`)
        : err;
    }
  }
}

// ── Convenience helper ────────────────────────────────────────────────────────

/**
 * Wrap a CrewAI HTTP service in a single call.
 *
 * @example
 * ```ts
 * const agent = await wrapCrewAI({
 *   agentId:    'borgkit://agent/writer',
 *   name:       'WriterCrew',
 *   version:    '1.0.0',
 *   owner:      '0xYourWallet',
 *   serviceUrl: 'http://localhost:8000',
 * });
 * await agent.registerDiscovery();
 * ```
 */
export async function wrapCrewAI(
  config: CrewAIPluginConfig,
  fetchCaps = true,
): Promise<ReturnType<CrewAIPlugin['wrap']>> {
  const plugin = new CrewAIPlugin(config);
  if (fetchCaps) {
    await plugin.fetchCapabilities().catch(() => { /* fall back to 'invoke' */ });
  }
  return plugin.wrap(plugin.createService());
}

File: templates/python/plugins/crewai_plugin.py
"""
CrewAIPlugin — Borgkit adapter for CrewAI agents.

Wraps a CrewAI Agent so it becomes fully discoverable and callable on the
Borgkit mesh without any changes to the original agent code.

How it works
────────────
1. At wrap time, capabilities are extracted from the agent's tool list.
   Each @tool-decorated function becomes one Borgkit capability.

2. When an AgentRequest arrives, translate_request() maps the Borgkit
   payload into a CrewAI Task description.

3. invoke_native() runs the task using a single-agent Crew, capturing
   the CrewOutput result.

4. translate_response() maps the CrewOutput back to an AgentResponse.

Install deps:
    pip install crewai crewai-tools

Usage:
    from plugins.crewai_plugin import wrap_crewai

    @tool("web_search")
    def web_search(query: str) -> str:
        \"\"\"Search the web for recent information.\"\"\"
        return f"Results for: {query}"

    agent = CrewAgent(
        role="Researcher",
        goal="Find accurate information",
        backstory="Expert research assistant.",
        tools=[web_search],
    )

    borgkit_agent = wrap_crewai(
        agent    = agent,
        name     = "ResearchAgent",
        agent_id = "borgkit://agent/researcher",
        owner    = "0xYourWallet",
        tags     = ["research", "crewai"],
    )
"""

from __future__ import annotations

from typing import Any, List, Optional

from plugins.base import (
    BorgkitPlugin,
    PluginConfig,
    CapabilityDescriptor,
    WrappedAgent,
)
from interfaces.agent_request  import AgentRequest
from interfaces.agent_response import AgentResponse


# ── Optional CrewAI import (soft dependency) ──────────────────────────────────

try:
    from crewai import Agent as CrewAgent, Task, Crew, Process
    from crewai.tools import BaseTool
    _CREWAI_OK = True
except ImportError:
    _CREWAI_OK = False
    CrewAgent = Any   # type: ignore[assignment,misc]


# ── Plugin config ─────────────────────────────────────────────────────────────

class CrewAIPluginConfig(PluginConfig):
    """
    Configuration for CrewAIPlugin.

    Inherits all fields from PluginConfig and adds CrewAI-specific options.
    """

    # CrewAI process type for running tasks.
    # Process.sequential runs tasks one at a time (default; safest for single-task use).
    process: str = "sequential"

    # Maximum number of iterations the crew is allowed per task.
    max_iter: int = 5

    # If True, the crew will emit verbose step-by-step output.
    verbose: bool = False

    # Optional system prompt prefix injected into every task description.
    system_prompt: Optional[str] = None


# ── Plugin ────────────────────────────────────────────────────────────────────

class CrewAIPlugin(BorgkitPlugin):
    """
    Borgkit ↔ CrewAI bridge.

    Each CrewAI tool decorated with @tool becomes one Borgkit capability.
    Borgkit AgentRequests are translated into CrewAI Tasks and run via a
    single-agent Crew.  Results are mapped back to AgentResponse.
    """

    def __init__(self, config: CrewAIPluginConfig):
        if not _CREWAI_OK:
            raise ImportError(
                "crewai is not installed — run: pip install crewai crewai-tools"
            )
        super().__init__(config)
        self._cfg: CrewAIPluginConfig = config

    # ── BorgkitPlugin abstract methods ────────────────────────────────────────

    def extract_capabilities(self, agent: CrewAgent) -> List[CapabilityDescriptor]:
        """
        Extract capabilities from the agent's tool list.

        Supports tools created with @tool decorator (crewai.tools.tool) and
        subclasses of BaseTool.
        """
        caps: List[CapabilityDescriptor] = []
        for t in (agent.tools or []):
            name = self._tool_name(t)
            desc = self._tool_description(t)
            params = self._tool_params(t)
            caps.append(CapabilityDescriptor(
                name        = name,
                description = desc,
                parameters  = params,
            ))
        return caps

    def translate_request(self, req: AgentRequest, agent: CrewAgent) -> dict:
        """
        Map a Borgkit AgentRequest to a CrewAI task description dict.

        The task description is built from:
        - A system prompt prefix (if configured)
        - The capability name
        - The full JSON payload
        """
        payload_str = _payload_to_string(req.payload)

        # Prefer an explicit 'task' or 'query' key in the payload
        task_body = (
            req.payload.get("task")
            or req.payload.get("query")
            or req.payload.get("input")
            or payload_str
        )

        prefix = self._cfg.system_prompt or f"You are performing the '{req.capability}' capability."
        description = f"{prefix}\n\nTask: {task_body}"

        return {
            "description":      description,
            "capability":       req.capability,
            "payload":          req.payload,
            "expected_output":  f"A complete result for the '{req.capability}' request.",
        }

    def invoke_native(self, agent: CrewAgent, translated: dict) -> Any:
        """
        Run the task using a single-agent Crew and return the raw CrewOutput.

        Each AgentRequest gets its own Crew instance.  CrewAI's shared state
        is reset between calls so there is no cross-request contamination.
        """
        task = Task(
            description     = translated["description"],
            agent           = agent,
            expected_output = translated["expected_output"],
        )

        crew = Crew(
            agents  = [agent],
            tasks   = [task],
            process = Process.sequential,
            verbose = self._cfg.verbose,
        )

        result = crew.kickoff()
        return result

    def translate_response(self, native_output: Any, req: AgentRequest) -> AgentResponse:
        """
        Map a CrewOutput (or plain string) back to a Borgkit AgentResponse.
        """
        # CrewOutput has .raw (str) and .tasks_output (list) attributes
        if hasattr(native_output, "raw"):
            content = native_output.raw
        else:
            content = str(native_output)

        return AgentResponse.success(req.request_id, {"content": content})

    # ── Tool introspection helpers ────────────────────────────────────────────

    @staticmethod
    def _tool_name(tool: Any) -> str:
        """Extract a clean capability name from a crewai tool."""
        # @tool-decorated functions expose .name
        if hasattr(tool, "name"):
            return str(tool.name)
        # BaseTool subclasses expose .name as a class attribute
        if hasattr(tool, "__class__") and hasattr(tool.__class__, "name"):
            return str(tool.__class__.name)
        return getattr(tool, "__name__", str(tool))

    @staticmethod
    def _tool_description(tool: Any) -> str:
        """Extract a human-readable description from a crewai tool."""
        if hasattr(tool, "description"):
            return str(tool.description)
        if hasattr(tool, "__doc__") and tool.__doc__:
            return tool.__doc__.strip()
        return ""

    @staticmethod
    def _tool_params(tool: Any) -> dict:
        """
        Best-effort extraction of parameter schema from a crewai tool.
        Returns a dict of {param_name: type_hint_string}.
        """
        import inspect
        params: dict = {}
        fn = getattr(tool, "_run", None) or getattr(tool, "func", None) or tool
        if callable(fn):
            try:
                sig = inspect.signature(fn)
                for pname, param in sig.parameters.items():
                    if pname in ("self", "cls"):
                        continue
                    ann = param.annotation
                    params[pname] = (
                        ann.__name__ if hasattr(ann, "__name__") else str(ann)
                    ) if ann is not inspect.Parameter.empty else "Any"
            except (ValueError, TypeError):
                pass
        return params


# ── Convenience wrapper ───────────────────────────────────────────────────────

def wrap_crewai(
    agent:    "CrewAgent",
    name:     str,
    agent_id: str,
    owner:    str,
    tags:     Optional[List[str]] = None,
    process:  str = "sequential",
    verbose:  bool = False,
    system_prompt: Optional[str] = None,
    **kwargs: Any,
) -> WrappedAgent:
    """
    Wrap a CrewAI Agent for the Borgkit mesh.

    After wrapping, the agent:
      - exposes each @tool-decorated function as a Borgkit capability
      - registers with LocalDiscovery (or whichever backend is configured)
      - handles AgentRequest / AgentResponse translation automatically
      - can be discovered by any other Borgkit agent via capability query

    Args:
        agent:         The CrewAI Agent instance to wrap.
        name:          Human-readable name shown in discovery results.
        agent_id:      Unique Borgkit URI, e.g. "borgkit://agent/researcher".
        owner:         Wallet or contract address of the agent owner.
        tags:          Optional list of search tags for discovery.
        process:       CrewAI Process type ("sequential" or "hierarchical").
        verbose:       Enable CrewAI verbose output.
        system_prompt: Optional prefix injected into every task description.

    Returns:
        A WrappedAgent that implements IAgent and is ready for registration.

    Example:
        from crewai import Agent as CrewAgent
        from crewai.tools import tool

        @tool("summarise")
        def summarise(text: str) -> str:
            \"\"\"Summarise a block of text into key points.\"\"\"
            return f"Summary: {text[:100]}..."

        agent = CrewAgent(
            role      = "Summariser",
            goal      = "Produce concise summaries",
            backstory = "Expert at distilling information.",
            tools     = [summarise],
        )

        wrapped = wrap_crewai(
            agent    = agent,
            name     = "SummaryAgent",
            agent_id = "borgkit://agent/summariser",
            owner    = "0xYourWallet",
            tags     = ["summarise", "nlp", "crewai"],
        )
    """
    cfg = CrewAIPluginConfig(
        name          = name,
        agent_id      = agent_id,
        owner         = owner,
        tags          = tags or [],
        process       = process,
        verbose       = verbose,
        system_prompt = system_prompt,
        **{k: v for k, v in kwargs.items()
           if k in ("version", "metadata_uri", "heartbeat_interval_s")},
    )
    plugin = CrewAIPlugin(cfg)
    return plugin.wrap(agent)


# ── Helpers ───────────────────────────────────────────────────────────────────

def _payload_to_string(payload: dict) -> str:
    """Convert an AgentRequest payload dict to a compact string for task descriptions."""
    import json
    try:
        return json.dumps(payload, ensure_ascii=False, indent=None)
    except (TypeError, ValueError):
        return str(payload)
